stdlib/public/core/GraphemeWalker.swift

```diff
@@ -0,0 +1,370 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2021 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+// GraphemeWalker is a tool to determine the Unicode grapheme break boundaries.
+// This works on strings of UTF8 and UTF16 (for bridged NSString) and returns
+// either the next or previous stride of the grapheme cluster. For example:
+//
+//   let someFlags = "🇺🇸🇬🇧"
+//
+// Byte and Scalar Representation:
+//
+//   |(       🇺       ) (        🇸       )|(        🇬       ) (        🇧      )|
+//   [240, 159, 135, 186, 240, 159, 135, 184, 240, 159, 135, 172, 240, 159, 135, 167]
+//    ^                                        ^
+//    | = String's start index                 | = String's next index
+//
+//
+// In this representation we see that there are a total of 16 utf8 code units,
+// but we don't want to report the count of a string as 16 to a user to sees
+// two emoji "character"s. We also see a total of 4 Unicode scalars, 2 of which
+// make up the single U.S. flag (two regional indicators), i.e. these two create
+// a "grapheme cluster" who represent a singular grapheme or in Swift's terms,
+// "Character". The job of GraphemeWalker is to determine these grapheme
+// cluster boundaries and to correctly report that the string "🇺🇸🇬🇧" has two
+// characters (🇺🇸 and 🇬🇧).
+extension Unicode {
+  internal struct _GraphemeWalker {
+    let scalars: String.UnicodeScalarView
+
+    init(_ scalars: String.UnicodeScalarView) {
+      self.scalars = scalars
+    }
+  }
+}
+
+extension Unicode._GraphemeWalker {
+  internal struct State {
+    var index: String.Index
+
+    var isBackwards: Bool = false
+    var isInEmojiSequence: Bool = false
+    var shouldBreakRI: Bool = false
+
+    static func forward(_ index: String.Index) -> State {
+      State(index: index)
+    }
+
+    static func backward(_ index: String.Index) -> State {
+      State(index: index, isBackwards: true)
+    }
+  }
+}
+
+extension Unicode._GraphemeWalker {
+  func advanceScalar(_ state: inout State) -> Unicode.Scalar? {
+    guard state.index != scalars.endIndex else {
+      return nil
+    }
+
+    let scalar = scalars[state.index]
+    scalars.formIndex(after: &state.index)
+    
+    return scalar
+  }
+
+  func retreatScalar(_ state: inout State) -> (Unicode.Scalar, String.Index)? {
```

Comment:
Stick a tuple label on the return value to make it clear what that index is. Is it the start or end of the returned scalar? How is it different to an updated state?